File: modules/aria-native/android/src/main/java/expo/modules/arianative/AriaNativeModule.kt
package expo.modules.arianative

import android.content.Context
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.security.SecureRandom
import java.security.cert.X509Certificate
import javax.net.ssl.SSLContext
import javax.net.ssl.X509TrustManager

/**
 * Bridges the JS setup screen to the same "aria_prefs" SharedPreferences
 * file that the pure-native call screens (AriaMessagingService,
 * IncomingCallActivity, InCallActivity -- injected separately via the
 * withAriaNative config plugin, living in com.aria.companion) read the
 * server URL from. Deliberately self-contained rather than importing
 * those classes: this module compiles as its own Gradle module, and
 * :app depends on it (not the other way around), so it can't reach into
 * com.aria.companion's code -- but SharedPreferences is scoped by file
 * name within the app's process, not by Kotlin package/Gradle module,
 * so both sides land on the exact same file on disk regardless.
 */
private const val PREFS = "aria_prefs"
private const val KEY_SERVER_URL = "server_url"

class AriaNativeModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("AriaNative")

    AsyncFunction("saveServerUrl") { url: String ->
      prefs().edit().putString(KEY_SERVER_URL, url.trimEnd('/')).apply()
    }

    AsyncFunction("getServerUrl") {
      prefs().getString(KEY_SERVER_URL, null)
    }

    AsyncFunction("registerDevice") { serverUrl: String, token: String ->
      registerDeviceBlocking(serverUrl, token)
    }
  }

  private fun prefs() =
    (appContext.reactContext ?: throw IllegalStateException("No Android context available"))
      .getSharedPreferences(PREFS, Context.MODE_PRIVATE)

  /** call_server.py uses a self-signed cert (see phase1/certs.py) -- fine
   * for our own local server, needs an explicit trust-all client since
   * this is plain OkHttp, not a WebView with an SSL-error override. */
  private fun registerDeviceBlocking(serverUrl: String, token: String): String {
    val trustManager = object : X509TrustManager {
      override fun checkClientTrusted(chain: Array<out X509Certificate>?, authType: String?) {}
      override fun checkServerTrusted(chain: Array<out X509Certificate>?, authType: String?) {}
      override fun getAcceptedIssuers(): Array<X509Certificate> = arrayOf()
    }
    val sslContext = SSLContext.getInstance("TLS")
    sslContext.init(null, arrayOf(trustManager), SecureRandom())
    val client = OkHttpClient.Builder()
      .sslSocketFactory(sslContext.socketFactory, trustManager)
      .hostnameVerifier { _, _ -> true }
      .build()

    val body = JSONObject().put("token", token).toString()
      .toRequestBody("application/json".toMediaType())
    val request = Request.Builder().url("$serverUrl/register_device").post(body).build()

    client.newCall(request).execute().use { response ->
      if (!response.isSuccessful) {
        throw Exception("Server responded with ${response.code}")
      }
      return "ok"
    }
  }
}

File: phone_app/modules/aria-native/src/AriaNativeModule.ts
import { NativeModule, requireNativeModule } from 'expo';

declare class AriaNativeModule extends NativeModule<{}> {
  /** Persists the server URL into the SharedPreferences file that the
   * native call/FCM code (com.aria.companion) reads from. */
  saveServerUrl(url: string): Promise<void>;
  getServerUrl(): Promise<string | null>;
  /** POSTs the FCM token to call_server.py's /register_device, trusting
   * its self-signed cert (this is our own local server). */
  registerDevice(serverUrl: string, token: string): Promise<string>;
  /** Opens the native in-call screen (which auto-starts a call), so you
   * can ring her yourself instead of only ever receiving. */
  openCallScreen(): Promise<void>;
  /** GET /settings -- the server's current settings, as a JSON string. */
  getSettings(serverUrl: string): Promise<string>;
  /** POST /settings with a JSON string (partial updates are fine). Returns
   * the saved settings as the server validated/clamped them. */
  saveSettings(serverUrl: string, json: string): Promise<string>;
  /** Any other server call; body is a JSON string, or '' for none. */
  request(serverUrl: string, method: string, path: string, body: string): Promise<string>;
  /** Saves her photo on the phone and uploads it to the server. */
  uploadAvatar(serverUrl: string, base64: string): Promise<string>;
  /** Refreshes the phone's copy of her photo; true if one exists. */
  cacheAvatar(serverUrl: string): Promise<boolean>;
  /** file:// URI of the phone's copy of her photo, or null. */
  avatarUri(): Promise<string | null>;
}

export default requireNativeModule<AriaNativeModule>('AriaNative');
